sdcm/mgmt.py

```diff
@@ -4,10 +4,292 @@
+            raise ScyllaManagerError("Encountered an error on '{}' command response".format(cmd))
+        lines = res.stdout.split('\n')
+        progress = 'N/A'
+        for line in lines:
+            if "Progress" in line:
+                progress = line.split()[3]
+                break
+        logger.debug("Task: {} progress is: {}".format(self.id,progress))
+        return progress
+
+    def is_status_in_list(self, list_status, check_task_progress=False):
+        """
+        Check if the status of a given task is in list
+        :param list_status:
+        :return:
+        """
+        status = self.status
+        if check_task_progress and status != TaskStatus.NEW: # check progress for all statuses except 'NEW'
+            progress = self.progress
+        return self.status in list_status
+
+    def wait_for_status(self, list_status, check_task_progress=True):
+        text = "Waiting until task: {} reaches status of: {}".format(self.id, list_status)
+        is_status_reached = False
+        is_status_reached = wait.wait_for(func=self.is_status_in_list, step=60,
+                                          text=text, list_status=list_status, check_task_progress=check_task_progress, timeout=3600)
+        return is_status_reached
+
+    def wait_for_status_done(self):
+        """
+        Wait and report that task is eventually reaching a 'final' status. meaning one of: done/error/stopped
+        :return:
+        """
+        logger.debug("Waiting for task: {} to be done..".format(self.id))
+        cur_status = self.status
+        if cur_status == TaskStatus.DONE:
+            return True
+        if cur_status in [TaskStatus.ERROR, TaskStatus.STOPPED]:
+            return False
+
+        if cur_status == TaskStatus.NEW:
+            logger.debug("Waiting for task: {} to start..".format(self.id))
+            list_status = [TaskStatus.RUNNING, TaskStatus.ERROR, TaskStatus.DONE]
+            res = self.wait_for_status(list_status=list_status)
+            if not res:
+                raise ScyllaManagerError("Unexpected result on waiting for task {} status".format(self.id))
+            cur_status = self.status
+            if cur_status == TaskStatus.ERROR:
+                return False
+
+        if cur_status == TaskStatus.RUNNING:
+            logger.debug("Waiting for task: {} to finish running..".format(self.id))
+            list_status = [TaskStatus.DONE, TaskStatus.ERROR]
+            res = self.wait_for_status(list_status=list_status)
+            if not res:
+                raise ScyllaManagerError("Unexpected result on waiting for task {} status {}".format(self.id, list_status))
+            cur_status = self.status
+            if cur_status == TaskStatus.ERROR:
+                return False
+        final_state = self.status
+        logger.debug("Task: {} final state is: {}".format(self.id, str(final_state)))
+        return self.status == TaskStatus.DONE
```

Comment:
If cur_status will be something else than what you expect here (NEW, RUNNING, ERROR, DONE, STOPPED) and you can't assume it'll always be these, your function isn't safe.

Reply:
if a returned status value is not on the list it will be caught by exception:
raise ScyllaManagerError("Could not recognize returned task status: {}".format(output_str))

Reply:
Still not safe, it'll do it only if the status isn't known.
However, if in the future there will be a new status, someone may add it to the ENUM and it will be recognized, but this function may miss it.
In general, not safe to handle if if if without any else.

Reply:
related code was changed/removed